### PR TITLE
Feat/distil

### PR DESCRIPTION
**Pull request by nhantien**
# Knowledge Distillation for YOLOv8 Wildfire Detection Model

## Overview
This PR implements knowledge distillation for a YOLOv8-based wildfire detection model. The implementation transfers knowledge from a pre-trained teacher model to a smaller student model while maintaining detection accuracy.

## Key Features
- **Model Architecture**: Uses YOLOv8n as the student model with 5 classes (FireBSI, LightningBSI, PersonBSI, SmokeBSI, VehicleBSI)
- **Distillation Strategy**: Combines detection loss with distillation loss using:
  - Complete IoU loss for bounding box regression
  - KL divergence for classification
  - Feature-level distillation from teacher to student
- **Training Pipeline**:
  - Configurable backbone freezing (default: 10 layers)
  - Automatic device detection (CPU/MPS/CUDA)
  - Checkpoint saving every N epochs
  - Progress tracking with tqdm
  - Comprehensive loss monitoring (bbox, classification, DFL, distillation)

## Technical Details
- Implements custom loss functions for distillation
- Uses non-max suppression for teacher predictions
- Supports both training and validation datasets
- Configurable through pipeline_config.json
- Modular design with separate functions for model loading, dataset preparation, and training

## Usage

Ensure you have `cd` into `src/distillation`

```python distil.py
```

## Dependencies
- PyTorch
- Ultralytics YOLOv8
- NumPy
- tqdm

### CODE REVIEW

**Comment by nhantien** [comment]
## Enhancements:
* Gradient clipping to mitigate exploding gradient (which might introduce NaN)
* Weighted sums of different loss instead of regular sum
* Added learning rate scheduler similar to how Ultralytics does it for Yolov8
